File: react-native/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/SPAnalyticsManager.kt
package com.storylyplacementreactnative.common

import android.content.Context
import android.util.Log
import com.appsamurai.storyly.analytics.STRAnalytics
import com.appsamurai.storyly.analytics.config.STRAnalyticsConfig
import com.appsamurai.storyly.analytics.model.STRAnalyticProduct
import com.appsamurai.storyly.analytics.model.STRAnalyticProductEvent
import com.storylyplacementreactnative.common.data.util.decodeFromJson

object SPAnalyticsManager {

    fun initialize(context: Context, configJson: String) {
        val map = decodeFromJson(configJson) ?: run {
            Log.e("[SPAnalyticsManager]", "Failed to parse analytics config JSON")
            return
        }
        val config = decodeSTRAnalyticsConfig(map) ?: run {
            Log.e("[SPAnalyticsManager]", "A valid analytics config is required")
            return
        }
        STRAnalytics.initialize(context.applicationContext, config)
    }

    fun track(eventJson: String) {
        val map = decodeFromJson(eventJson) ?: return
        val event = decodeSTRAnalyticProductEvent(map["event"] as? String) ?: return
        val products = (map["products"] as? List<Map<String, Any?>>)
            ?.mapNotNull { decodeSTRAnalyticProduct(it) } ?: return
        STRAnalytics.track(event, products)
    }
}

internal fun decodeSTRAnalyticsConfig(config: Map<String, Any?>): STRAnalyticsConfig? {
    val token = config["token"] as? String ?: return null
    val builder = STRAnalyticsConfig.Builder()
    (config["userId"] as? String)?.let { builder.setUserId(it) }
    return builder.build(token)
}

internal fun decodeSTRAnalyticProductEvent(event: String?): STRAnalyticProductEvent? {
    event ?: return null
    return STRAnalyticProductEvent.entries.firstOrNull { it.name == event }
}

internal fun decodeSTRAnalyticProduct(product: Map<String, Any?>?): STRAnalyticProduct? {
    product ?: return null
    val productId = product["productId"] as? String ?: return null
    val productGroupId = product["productGroupId"] as? String ?: return null
    val title = product["title"] as? String ?: return null
    val price = (product["price"] as? Number)?.toFloat() ?: return null
    return STRAnalyticProduct(
        productId = productId,
        productGroupId = productGroupId,
        title = title,
        desc = product["desc"] as? String,
        price = price,
        salesPrice = (product["salesPrice"] as? Number)?.toFloat(),
        quantity = (product["quantity"] as? Number)?.toInt() ?: 1,
    )
}

File: react-native/storyly-placement-react-native/ios/Common/SPAnalyticsManager.swift
import Foundation
import StorylyAnalytics

@objc public class SPAnalyticsManager: NSObject {

    @objc public static let shared = SPAnalyticsManager()

    private override init() {
        super.init()
    }

    @objc public func initialize(configJson: String) {
        guard let dict = decodeFromJson(configJson),
              let config = decodeSTRAnalyticsConfig(dict) else {
            print("[SPAnalyticsManager] A valid analytics config is required")
            return
        }
        STRAnalytics.initialize(config: config)
    }

    @objc public func track(eventJson: String) {
        guard let dict = decodeFromJson(eventJson),
              let event = decodeSTRAnalyticProductEvent(dict["event"] as? String),
              let productsArray = dict["products"] as? [[String: Any]] else {
            return
        }
        let products = productsArray.compactMap { decodeSTRAnalyticProduct($0) }
        STRAnalytics.track(productEvent: event, products: products)
    }
}

func decodeSTRAnalyticsConfig(_ dict: [String: Any]) -> STRAnalyticsConfig? {
    guard let token = dict["token"] as? String else { return nil }

    let builder = STRAnalyticsConfig.Builder()
    if let userId = dict["userId"] as? String {
        _ = builder.setUserId(userId: userId)
    }
    return builder.build(token: token)
}

func decodeSTRAnalyticProductEvent(_ event: String?) -> STRAnalyticProductEvent? {
    guard let event = event else { return nil }
    return STRAnalyticProductEvent(value: event)
}

func decodeSTRAnalyticProduct(_ dict: [String: Any]) -> STRAnalyticProduct? {
    guard let productId = dict["productId"] as? String,
          let productGroupId = dict["productGroupId"] as? String,
          let title = dict["title"] as? String,
          let price = dict["price"] as? Double else {
        return nil
    }

    let desc = dict["desc"] as? String
    let salesPrice = dict["salesPrice"] as? Double
    let quantity = dict["quantity"] as? Int ?? 1

    return STRAnalyticProduct(
        productId: productId,
        productGroupId: productGroupId,
        title: title,
        desc: desc,
        price: Float(price),
        salesPrice: salesPrice as NSNumber?,
        quantity: quantity
    )
}
